Bug 1125835 - Error pages

wesj opened the pull request:
This uses a little trickery for errors pages using the local server. Essentially it maintains a list of explicitly requested loads. If the page is in that list, the error page is shown (and then its removed from the list). If a page isn't in the list, we pull the original url out of its url and redirect to it.

To handle back and forward navigation, we also listen for "popstate" events in the document. If they happen, we know that we're being shown from the cache via back-foward navigation and we also force a reload of the page (which will force a redirect via the mechanism I talked about above).

I'm not using a ton of the info WkWebView gives us here beyond a localized description. We could bring in more later. I made "actions" injectable into the page so that we could potentially add special ones for special cases. The HTML/CSS is a heavily redacted and modified version of Android/Desktop's.

st3fan commented on Client/Frontend/Browser/ErrorPageHelper.swift:
```diff
@@ -0,0 +1,171 @@
+import WebKit
+import Shared
+
+private let server: GCDWebServer = GCDWebServer()
```
Why is this here? Aren't you using the server that is created in `AppDelegate`?

st3fan commented:
Looks pretty good. I'll give this a try tomorrow morning.

thebnich commented on Client/Frontend/Browser/ErrorPageHelper.swift:
```diff
@@ -0,0 +1,171 @@
+//
+//  Created by Wes Johnston on 5/13/15.
+//  Copyright (c) 2015 Mozilla. All rights reserved.
+//
```
driveby ^

wesj commented:
Fixed up the comments. Also added some code to keep these out of history. Sadly, that has to be in BrowserViewController. Putting it in Storage means making Storage aware of the Error Pages, which also required making it aware of the WebServer. Trying to put it in Profile also failed (once you make the Profile know about WebServer the project won't build (because Profile is used in Extensions and Extensions have no AppDelegate which GCDWebServer requires).

wesj commented:
lttm

fxios-buildasaur commented:
Result of integration 1
Integration took 10 minutes and 38 seconds.
All 124 tests passed, but please fix 14 warnings.
